File: apps/quickink/ios/QuickInk/Stories/StoryNotePickerSheet.swift
/*
 * StoryNotePickerSheet.swift
 *
 * Stories Phase 2 follow-up — single-select picker over the user's
 * notepad entries (the daily-journal table from ReleafCoreNotes).
 * Opens from the "Choose a note" row in `StoryAddSheet` and returns
 * the picked entry id to the caller. The editor then inserts a
 * `story_item` of `kind = .note` with `refId = entryId`.
 *
 * Visual treatment differs from `StoryLibraryPickerSheet`: no
 * thumbnail (notes are text), so the row shows title (or "Daily
 * note") plus a 60-char excerpt of the body plus the entry date.
 *
 * Mirror of Android `StoryNotePickerSheet.kt`.
 */

import GRDB
import SwiftUI

struct StoryNotePickerSheet: View {

    let userId: String
    var onPick: (_ entryId: String) -> Void
    var onCancel: () -> Void

    @State private var rows: [PickerRow] = []
    @State private var loading: Bool = true

    var body: some View {
        VStack(spacing: 0) {
            handle
            header
            content
        }
        .background(QuickInkColors.surface)
        .task { await load() }
    }

    private var handle: some View {
        Capsule()
            .fill(QuickInkColors.border)
            .frame(width: 38, height: 4)
            .padding(.top, QuickInkSpacing.s2)
            .padding(.bottom, QuickInkSpacing.s3)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("Choose a note")
                .font(QuickInkText.editorial)
                .foregroundStyle(QuickInkColors.ink)
            Text("Daily notes from your journal.")
                .font(QuickInkText.bodyItalic)
                .foregroundStyle(QuickInkColors.inkSoft)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, QuickInkSpacing.s4)
        .padding(.bottom, QuickInkSpacing.s3)
    }

    @ViewBuilder
    private var content: some View {
        if loading {
            HStack {
                Spacer()
                ProgressView().padding(.vertical, QuickInkSpacing.s5)
                Spacer()
            }
        } else if rows.isEmpty {
            Text("No notes yet. Write one in the notepad first.")
                .font(QuickInkText.bodyItalic)
                .foregroundStyle(QuickInkColors.inkSoft)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.bottom, QuickInkSpacing.s4)
        } else {
            ScrollView(showsIndicators: false) {
                VStack(spacing: QuickInkSpacing.s2) {
                    ForEach(rows, id: \.id) { row in
                        Button(action: { onPick(row.id) }) {
                            pickerRow(row)
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.bottom, QuickInkSpacing.s6)
            }
        }
    }

    private func pickerRow(_ row: PickerRow) -> some View {
        HStack(alignment: .top, spacing: QuickInkSpacing.s3) {
            VStack(alignment: .leading, spacing: 4) {
                Text(row.title)
                    .font(QuickInkText.editorial)
                    .foregroundStyle(QuickInkColors.ink)
                    .lineLimit(1)
                if !row.excerpt.isEmpty {
                    Text(row.excerpt)
                        .font(QuickInkFont.serif(12, weight: .regular, italic: true))
                        .foregroundStyle(QuickInkColors.inkSoft)
                        .lineLimit(2)
                }
                Text(row.entryDate)
                    .font(.system(size: 11))
                    .foregroundStyle(QuickInkColors.muted)
            }
            Spacer(minLength: 0)
        }
        .padding(QuickInkSpacing.s2 + 2)
        .background(QuickInkColors.bg)
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                .strokeBorder(QuickInkColors.border, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
        .frame(maxWidth: .infinity, alignment: .leading)
        .contentShape(Rectangle())
    }

    // MARK: - Data

    private struct PickerRow: Identifiable {
        let id: String
        let title: String
        let excerpt: String
        let entryDate: String
    }

    @MainActor
    private func load() async {
        loading = true
        let userId = self.userId
        let queue = QuickInkDatabase.shared.dbQueue
        let fetched: [PickerRow] = (try? await queue.read { db -> [PickerRow] in
            let rows = try Row.fetchAll(db, sql: """
                SELECT id, title, notes, entry_date
                FROM notepad_entries
                WHERE user_id = ? AND deleted_at IS NULL
                ORDER BY entry_date DESC
                LIMIT 100
                """, arguments: [userId])
            return rows.map { row -> PickerRow in
                let body = (row["notes"] as String?) ?? ""
                let title = ((row["title"] as String?)?.trimmingCharacters(in: .whitespacesAndNewlines))
                    .flatMap { $0.isEmpty ? nil : $0 } ?? "Daily note"
                let excerpt: String = {
                    let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
                    if trimmed.isEmpty { return "" }
                    let cut = String(trimmed.prefix(60))
                    return trimmed.count > 60 ? cut + "…" : cut
                }()
                return PickerRow(
                    id:        row["id"],
                    title:     title,
                    excerpt:   excerpt,
                    entryDate: row["entry_date"]
                )
            }
        }) ?? []
        self.rows = fetched
        self.loading = false
    }
}

File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/stories/StoryNotePickerSheet.kt
/*
 * StoryNotePickerSheet.kt
 *
 * Stories Phase 2 follow-up — single-select picker over the user's
 * notepad entries. Mirror of iOS `StoryNotePickerSheet.swift`.
 */

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.quickink.mobile.features.stories

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.QuickInkApp
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing
import kotlinx.coroutines.flow.first

@Composable
fun StoryNotePickerSheet(
    userId: String,
    onPick: (entryId: String) -> Unit,
    onDismiss: () -> Unit,
) {
    val colors  = LocalQuickInkColors.current
    val type    = LocalQuickInkTypography.current
    val context = LocalContext.current
    val app     = remember(context) { context.applicationContext as QuickInkApp }

    var rows by remember { mutableStateOf<List<PickerRow>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }

    LaunchedEffect(userId) {
        val entries = app.database.notepadDao().observeActive(userId).first()
        rows = entries.take(100).map { entry ->
            val body = entry.notes.trim()
            val excerpt = if (body.isEmpty()) "" else
                if (body.length > 60) body.take(60) + "…" else body
            val title = entry.title?.trim()?.takeIf { it.isNotEmpty() } ?: "Daily note"
            PickerRow(
                id        = entry.id,
                title     = title,
                excerpt   = excerpt,
                entryDate = entry.entryDate,
            )
        }
        loading = false
    }

    val sheetState = rememberModalBottomSheetState()
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState       = sheetState,
        containerColor   = colors.surface,
    ) {
        Column(
            modifier = Modifier.padding(
                start  = QuickInkSpacing.s4,
                end    = QuickInkSpacing.s4,
                bottom = QuickInkSpacing.s6,
            ),
        ) {
            Text(text = "Choose a note", style = type.editorial, color = colors.ink)
            Text(
                text     = "Daily notes from your journal.",
                style    = type.bodyItalic,
                color    = colors.inkSoft,
                modifier = Modifier.padding(bottom = QuickInkSpacing.s3),
            )
            when {
                loading -> {
                    Box(
                        modifier = Modifier.fillMaxWidth().padding(vertical = QuickInkSpacing.s5),
                        contentAlignment = Alignment.Center,
                    ) {
                        CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(20.dp))
                    }
                }
                rows.isEmpty() -> {
                    Text(
                        text  = "No notes yet. Write one in the notepad first.",
                        style = type.bodyItalic,
                        color = colors.inkSoft,
                    )
                }
                else -> {
                    LazyColumn(
                        contentPadding = PaddingValues(bottom = QuickInkSpacing.s3),
                        verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s2),
                        modifier = Modifier.height(400.dp),
                    ) {
                        items(rows.size) { idx ->
                            val row = rows[idx]
                            PickerRowView(row = row, onClick = { onPick(row.id) })
                        }
                    }
                }
            }
        }
    }
}

private data class PickerRow(
    val id: String,
    val title: String,
    val excerpt: String,
    val entryDate: String,
)

@Composable
private fun PickerRowView(row: PickerRow, onClick: () -> Unit) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(QuickInkRadius.md))
            .background(colors.bg)
            .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.md))
            .clickable(onClick = onClick)
            .padding(QuickInkSpacing.s2 + 2.dp),
    ) {
        Text(
            text     = row.title,
            style    = type.editorial,
            color    = colors.ink,
            maxLines = 1,
        )
        if (row.excerpt.isNotEmpty()) {
            Text(
                text      = row.excerpt,
                style     = type.bodyItalic,
                color     = colors.inkSoft,
                fontSize  = 12.sp,
                fontStyle = FontStyle.Italic,
                maxLines  = 2,
                modifier  = Modifier.padding(top = 4.dp),
            )
        }
        Text(
            text     = row.entryDate,
            color    = colors.muted,
            fontSize = 11.sp,
            modifier = Modifier.padding(top = 4.dp),
        )
    }
}
